Fixes #12720 - Add tabs for puppet on proxy show page

tbrisker opened the pull request.

tbrisker committed:
Fixes #12720 - Add tabs for puppet on proxy show page

This patch adds two tabs to the proxy show page related to puppet
1. Puppet - this tab is for proxies managing a puppet master. This tab
displays information about the puppet master, such as number of classes
in each environment, latest reports and so on.
2. Puppet CA - This tab incorporates the two seperate pages that were
already present for managing the puppet ca's certificates and autosign
entry. It also incorporated the new autosign form in a modal view.

ohadlevy commented on app/controllers/puppetca_controller.rb:
```diff
@@ -3,18 +3,18 @@ def index
     Rails.cache.delete("ca_#{@proxy.id}") if params[:expire_cache] == "true"
-    certs = find_certs
-    @certificates = certs.sort.paginate :page => params[:page], :per_page => Setting::General.entries_per_page
+    @certificates = SmartProxies::PuppetCA.all(@proxy).sort
```
what happens to filtering and default pending state?

tbrisker replied:
all handled by datatables in JS

ohadlevy replied:
i guess we can drop the sort then?

tbrisker replied:
removed

ohadlevy commented on app/helpers/smart_proxies_helper.rb:
```diff
@@ -3,12 +3,6 @@ module SmartProxiesHelper
 
   def proxy_actions(proxy, authorizer)
     [ display_link_if_authorized(_("Edit"), hash_for_edit_smart_proxy_path(:id => proxy), :class => 'edit_two_pane') ] +
-    [if proxy.has_feature?('Puppet CA')
```
this was useful for direct navigation, any chance to keep this kind of actions (one click vs multiple)

ohadlevy commented on app/views/smart_proxies/plugins/_puppet_ca.html.erb:
```diff
@@ -0,0 +1,12 @@
+  <%= Host::Managed.where(:puppet_ca_proxy_id => @smart_proxy.id).count %>
+</div>
+<% if authorized_via_my_scope(:smart_proxy, :view_smart_proxies_puppet_ca, @smart_proxy) %>
+  <div class="col-md-12" data-ajax-url="<%= smart_proxy_puppetca_index_path(:smart_proxy_id => @smart_proxy) %>" data-on-complete="onContentLoad"><%= spinner %></div>
```
I find it a bit confusing when its in one tab, mostly the fact the table has no fixed-size, so going over pagination is making my page "jump" around.

ohadlevy commented on app/views/smart_proxies/plugins/_puppet_ca.html.erb:
```diff
@@ -0,0 +1,12 @@
+  <div class="col-md-12" data-ajax-url="<%= smart_proxy_puppetca_index_path(:smart_proxy_id => @smart_proxy) %>" data-on-complete="onContentLoad"><%= spinner %></div>
+<% end %>
+<% if authorized_via_my_scope(:smart_proxy, :view_smart_proxies_autosign, @smart_proxy) %>
+  <div class="col-md-12" data-ajax-url="<%= smart_proxy_autosign_index_path(:smart_proxy_id => @smart_proxy) %>" data-on-complete="onContentLoad"><%= spinner %></div>
```
the autosign new link is not styled, plus, i wonder if it should be modal in this case vs leaving the page.

ohadlevy commented on app/views/smart_proxies/plugins/_puppet_ca.html.erb:
```diff
@@ -0,0 +1,12 @@
+<%= show_feature_version('puppetca') %>
+<div class="col-md-4"><%= _('Hosts managed') %></div>
+<div class="col-md-8">
+  <%= Host::Managed.where(:puppet_ca_proxy_id => @smart_proxy.id).count %>
```
I'm missing CA counters here, as I think that if you have pending certs it would be useful to see them here as a number.
this also means we are prefetching the data for these numbers, so maybe reconsider loading this whole cert related info in one ajax?

lzap commented:
Ok, taking the review. First, can we start providing some information? What is the goal here? All we have is this sentence in the ticket: "In the proxy's show page - I would like to know how many puppet classes are on that proxy". This doesn't even match what's being added here. Our community members do not attend our feature teams scrums and I am bit confused too.

lzap commented:
First of all, I would like to discuss the UI part:

![image](https://cloud.githubusercontent.com/assets/49752/12141508/5b782b4a-b470-11e5-8110-ee0dd1039d4c.png)
- I think we should split both tables into two separate tabs, this looks clunky. I don't believe this fits in one page with some real data.
- Why Puppet CA generic status information is on a separate one, I think it makes more sense to show generic information (Version, Host managed) on the Status together with others.
- The New button and Search box do not follow our UI style.

lzap commented:
Ok scratch my comment about "manual edits". Now after I read the code I understand you are actually _moving_ Puppet CA and Autosign entries from their own pages here.

ohadlevy commented:
@lzap @orrabin was "playing" around with the dhcp part, we talked about potentially merging foreman-dhcp plugin as well down the road.

lzap commented on app/views/smart_proxies/plugins/_dhcp.html.erb:
```diff
@@ -1,6 +1,6 @@
 <div class="row">
   <h3><%= _('DHCP') %></h3>
-  <%= render :partial => 'smart_proxies/plugins/plugin_version', :locals => {:feature => feature, :proxy => @smart_proxy} %>
+  <%= show_feature_version('dhcp') %>
```
Use the `feature` variable here.

lzap commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -69,5 +73,18 @@
+      <% end %>
+      </div>
+    <% end %>
+    </div>
```
Now I understand why you tried to squeeze everything into one tab. Well, I think we need to get rid of iterating though features in order to get a usable UI. Or find a better way.

lzap commented:
@ohadlevy yeah if we can put this here as an extra tab, it looks like it's not much work (following the same pattern as here). Big help for users struggling to delete DHCP/DNS records.

tbrisker commented:
Still WIP, current status:
- [x] Show certificates and autosign tables in tabs under puppet ca
- [x] Allow quick navigation to certs and autosign tables
- [x] Add charts on puppet page
- [x] Link host counts to relevant host index
- [x] Show new autosign form in modal
- [x] Allow filtering certificates
- [x] Fix styling for new autosign button 
- [x] Add counters for CA states
- [x] Add more info in commit message
- [x] Add puppet class counters (~~https://github.com/theforeman/smart-proxy/pull/356 could ease this~~ handled on the foreman side for now)

lzap commented:
Let me know when this is ready for review. FYI tests failed.

ohadlevy commented:
[test](not sure if its random failure)

ohadlevy commented:
@tbrisker is this still WIP or ready for-review?

tbrisker commented:
Ready for review.

ohadlevy commented on app/assets/javascripts/proxy_status.js:
```diff
@@ -1,8 +1,12 @@
+  setTab();
 });
 
+$(window).on('hashchange', setTab); //needed to handle some title action buttons
```
why do we need window?

tbrisker replied:
Because the `hashchange` event that is fired when clicking on a link to an anchor on the same page (as is the case when selecting autosign or certificates from the drop down action) is fired on the window, not the document.

ohadlevy commented on app/assets/javascripts/proxy_status.js:
```diff
@@ -95,3 +103,29 @@ function populateData(response, item) {
+  if (anchor.length) {
+    var parent_tab = $(anchor).parents('.tab-pane');
+    if (parent_tab.exists()){
+      $('.nav-tabs a[href=#'+parent_tab[0].id+']').tab('show');
```
a comment on what this is would be nice?

tbrisker replied:
added

ohadlevy commented on app/assets/javascripts/proxy_status.js:
```diff
@@ -95,3 +103,29 @@ function populateData(response, item) {
+
+function certTable() {
+  activateDatatables();
+  filterCerts('valid');
```
could this define via a parameter? so then if you have a pending cert link we could reuse the same table?

ohadlevy replied:
or in other words, can you add a counter for how many pending certs are there, and if you click on it redirects you to the list ? :)

tbrisker replied:
It is possible, but not trivial. I think that this could wait for a future PR if the need arises.
Note that right now `filterCerts` does accept a parameter, so once the table is loaded we can link to filter it (i.e. from a counter on the show page)

ohadlevy commented on app/controllers/puppetca_controller.rb:
```diff
@@ -33,15 +35,4 @@ def destroy
-    if params[:state].blank?
-      SmartProxies::PuppetCA.find_by_state(@proxy, "valid") +
-        SmartProxies::PuppetCA.find_by_state(@proxy, "pending")
-    elsif params[:state] == "all"
```
note that the original filter showed both valid and pending.

tbrisker replied:
fixed.

ohadlevy commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -45,6 +45,13 @@ def tftp_server
   end
 
+  def puppet_environments
+    envs = @smart_proxy.public_send(:environment_stats)
```
use  requested_data just like the ping method?

tbrisker replied:
`requested_data` renders a json response, here I want to render a partial with that data.

ohadlevy replied:
ok, can we change requested_data to allow that? it makes little sense to
use more than one method

tbrisker replied:
Actually this was just silly to use public_send here. I only need the data for the partial, `requested_data` gets data from the proxy and sends a json.

ohadlevy commented on app/helpers/dashboard_helper.rb:
```diff
@@ -89,7 +89,7 @@ def latest_events
 
   def translated_header(shortname, longname)
-    "<th><span class='small' title='' data-original-title='#{longname}'>#{shortname}</span></th>"
+    "<th width='50px' class='ca'><span class='small' title='' data-original-title='#{longname}'>#{shortname}</span></th>"
```
why here vs a css class?

tbrisker replied:
fixed

ohadlevy commented on app/views/autosign/_form.html.erb:
```diff
@@ -1,9 +1,24 @@
-    <%= label_tag(_("Name")) %>
-    <div class="input">
-      <%= text_field_tag(:id) %>
+<div class="modal fade" id="autosignModal" tabindex="-1" role="dialog" aria-labelledby="autosignModalLabel">
```
does it make sense to introduce a modal  helper?

tbrisker replied:
made it into a layout partial. too much nested html for a helper to be useful.

ohadlevy commented on app/views/autosign/_form.html.erb:
```diff
@@ -1,9 +1,24 @@
+  <div class="modal-dialog" role="document">
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
```
nitpick: can use modal_close helper

tbrisker replied:
fixed

ohadlevy commented on app/views/dashboard/_reports_widget.html.erb:
```diff
@@ -5,7 +5,7 @@
   <p class="ca"><%= _("No interesting reports received in the last week") %></p>
 <% else %>
-  <table class="table table-striped table-bordered ellipsis">
+  <table class="table table-striped table-bordered table-fixed">
```
should we use table_css_classes helper?

tbrisker replied:
fixed

ohadlevy commented on app/views/puppetca/_list.html.erb:
```diff
@@ -9,7 +12,7 @@
       <th><%= _("Expires") %></th>
       <th><%= _("Fingerprint") %></th>
-      <th><%= _('Actions') %></th>
+      <th></th>
```
why remove the actions? afaiu we were trying to add them to most pages?

tbrisker replied:
Looks like a mistake in the rebase. will revert.

ohadlevy commented on app/views/smart_proxies/plugins/_puppet.html.erb:
```diff
@@ -0,0 +1,29 @@
+<div class="tab-content proxy-content" id="proxy-puppet-tab-content">
+  <div class="tab-pane active in" id="puppet_dashboard">
+    <div class="col-md-5">
+      <%= render :partial => 'dashboard/reports_widget' %>
```
should these be refactored to accept a list of reports or hosts instead of setting @hosts and @reports?

tbrisker replied:
I think this makes sense, though I think there was a reason I used instance variables in the original dashboard... this can be refactored separately in any case, not to complicate this PR even more.

ohadlevy commented on config/environments/development.rb:
```diff
@@ -36,7 +36,7 @@
   # Adds additional error checking when serving assets at runtime.
   # Checks for improperly declared sprockets dependencies.
   # Raises helpful error messages.
-  config.assets.raise_runtime_errors = true
```
wrong?

tbrisker replied:
fixed

ohadlevy commented:
@tbrisker missing tests?

ohadlevy commented:
setting on WoC

tbrisker commented:
I think I've addressed all comments so far except the tests - the code can be reviewed while I add the tests.

lzap commented:
Removing myself as I am not working two days, looks like this is on the good track.

tbrisker commented on app/helpers/puppetca_helper.rb:
```diff
@@ -1,8 +1,19 @@
-    states = [N_('revoked'), N_('pending'), N_('valid')]
-    select_tag "Filter", options_for_select([''] + states.map { |s| [_(s), s] }, params[:state]), :class => "puppetca-filters",
-               :onchange => "window.location.href = '#{smart_proxy_puppetca_index_path(@proxy)}' + (this.value == '' ? '' : ('?state=' + this.value))"
+    select_tag "Filter", options_for_select([[_('valid or pending'), _('valid')+'|'+_('pending')]] +
```
since the filtering is done by datatables on the table data itself, the states will already be translated, so we need to have the filter use the translated strings as well, and not the original strings as was before.

lzap replied:
Just for the record, this needs to be styled, looks a bit plain.

tbrisker commented on app/helpers/layout_helper.rb:
```diff
@@ -469,8 +469,8 @@ def alert_actions(actions)
     end
   end
 
-  def modal_close(data_dismiss = 'modal', text = _('Close'))
```
this wasn't actually used anywhere, and makes no sense to have anything other then `modal` here.

tbrisker commented:
I've added a few tests, let me know if there is any area where more testing is needed - most of the code already existed and just changed layout.

tbrisker commented on app/services/proxy_status.rb:
```diff
@@ -26,6 +26,18 @@ def tftp_server
     end
   end
 
+  def environment_stats
```
I'm a bit concerned that these requests might take a long time and should be handled more quickly on the proxy using https://github.com/theforeman/smart-proxy/pull/356 as a start and improving on it. However if testing indicates that this runs in an acceptable time (a few seconds for a largish setup) that PR can be closed.

tbrisker replied:
tested on about 200 classes and 3 environments, this runs in about 2 seconds, so for now I closed the smart proxy pr, will reopen it if we see this is getting to be a performance bottleneck.

ohadlevy commented:
in the logs i see  | ActionController::RoutingError (No route matches [GET] "/assets/sort_both.png"):

ohadlevy commented on app/views/smart_proxies/plugins/_puppet_envs.html.erb:
```diff
@@ -0,0 +1,20 @@
+    <% envs.each do |name, count| %>
+      <tr>
+        <td class="ellipsis"><%= name %></td>
+        <td><%= count %></td>
```
why table? it could also be nice to show the numbers in large?

tbrisker replied:
I still haven't found a good way to display the numbers that clarifies they are the class count. Also, per our discussion, I think we might want to add columns here counting the number of new, updated and obsolete classes, and/or a link to the import page for the specific environment.

ohadlevy commented on app/views/autosign/_form.html.erb:
```diff
@@ -1,9 +1,13 @@
+                                                     :buttons => [ modal_close(_('Cancel')),
+                                                                   submit_tag(_('Save'), :class => "btn btn-success") ]} do %>
+    <div class="clearfix">
+      <%= label_tag(_("Name")) %>
```
cant you use the text helpers (as you have an "f" object)

tbrisker replied:
fixed

tbrisker replied:
Or, more precisely, no - since f here was fictive, as autosign is not an object.  But I made the layout more similar to our other forms.

ohadlevy commented on app/views/autosign/_form.html.erb:
```diff
@@ -1,9 +1,13 @@
+    <div class="clearfix">
+      <%= label_tag(_("Name")) %>
+      <div class="input">
+        <%= text_field_tag(:id) %>
```
input is too short, I would expend it.

tbrisker replied:
fixed

ohadlevy commented on app/models/smart_proxy.rb:
```diff
@@ -30,7 +30,7 @@ class SmartProxy < ActiveRecord::Base
   scoped_search :in => :features, :on => :name, :rename => :feature, :complete_value => :true
 
-  delegate :version, :tftp_server, :to => :proxy_status
+  delegate :version, :tftp_server, :environment_stats, :to => :proxy_status
```
can you please add a TODO as this needs to move to its own proxy status class (see #3069)

tbrisker replied:
I assume you meant the TODO in the proxy_status object, fixed

ohadlevy commented on app/services/proxy_status.rb:
```diff
@@ -26,6 +26,18 @@ def tftp_server
 
+  def environment_stats
+    raise ::Foreman::Exception.new(N_('No Puppet feature for %s'), proxy.to_label) unless proxy.has_feature?('Puppet')
+    Rails.cache.fetch("proxy_#{proxy.id}/environments", :expires_in => cache_duration) do
```
you dont handle cache deletion

tbrisker replied:
fixed

ohadlevy commented on app/helpers/puppetca_helper.rb:
```diff
@@ -1,8 +1,19 @@
+  def state_counters
+    STATES.map do |state|
+      content_tag(:div, _(state)+":", :class => 'col-md-2')+
+        content_tag(:div, @certificates.count{|c| c.state == state}.to_s, :class => 'col-md-10')
```
can you highlight pending if its > 0 ?

tbrisker replied:
Even better, made this into patternfly cards and changed the order so pending is the first displayed, I think it should be obvious enough.

ohadlevy commented on app/services/proxy_status.rb:
```diff
@@ -26,6 +26,18 @@ def tftp_server
+      fetch_proxy_data do
+        proxy = ProxyAPI::Puppet.new(:url => @proxy.url)
+        env_counts = {}
+        proxy.environments.each{ |env|  env_counts[env] = proxy.class_count(env) }
```
nitpick: extra whitespace

tbrisker commented:
The routing error for assets is unrelated, and has to do with some weirdness of the asset pipeline in production. Actually locally production does not show any assets (except for js and css) for me for some reason.

tbrisker commented:
I think I've addressed everything. I think this looks better now. I think I need some sleep.

ohadlevy commented on app/helpers/layout_helper.rb:
```diff
@@ -505,13 +505,17 @@ def new_child_fields_template(form_builder, association, options = { })
 
+  def card_status_mini(options)
+    options.reverse_merge! :count => '', :icon => '', :callback => '', :size => 'col-md-4', :notifications => ''
+    render :partial => 'common/card_status_mini', :locals => options
```
i think you forgot to checkin this file?

tbrisker replied:
That's what happens when pushing late at night :) fixed.

ohadlevy replied:
i find the cards a bit too big? they also dont change the drop down status.
![recorded](https://cloud.githubusercontent.com/assets/20167/12376917/53037dec-bd13-11e5-8aa6-5c2f803dc902.gif)

ohadlevy replied:
on second thought, maybe they just miss a background color (cards-pf css class) as shown at https://www.patternfly.org/patterns/aggregate-status-card/

ohadlevy replied:
unresolved yet.

tbrisker replied:
They change the dropdown now. 
Not sure about size, I could make them smaller but we would have extra white space on the side.
I tried adding the background and I think it looks really bad to have a grey area in the middle of the white page.

ohadlevy commented on app/helpers/autosign_helper.rb:
```diff
@@ -1,5 +1,5 @@
 module AutosignHelper
   def autosign_form
-    link_to "New", new_smart_proxy_autosign_path(@proxy) if authorized_for(:controller => "SmartProxies::Autosign", :action => :create, :auth_object => @proxy)
+    button_tag("New", :type => 'button', :class => 'btn btn-primary fr', :data => {:toggle => 'modal', :target =>'#autosignModal'}) if authorized_for(:controller => "SmartProxies::Autosign", :action => :create, :auth_object => @proxy)
```
i find the modal positioning too high in the screen (e.g it override the top bar and its very short)

ohadlevy replied:
also, when opening/closing the modal, there is a slight shift of some of the divs to the left

tbrisker replied:
Pushed it down a bit.
I didn't see any shifting when opening the modal, could you add a screen shot?

ohadlevy replied:
![recorded](https://cloud.githubusercontent.com/assets/20167/12376911/1fedf996-bd13-11e5-9c7b-442c4b4340db.gif)

dLobatog replied:
This has been fixed

ohadlevy replied:
@dLobatog also the minor "jump/shift"

tbrisker replied:
I can't seem to reproduce the shift, looks like a scroll bar is being added on the side? possibly something to do with a specific resolution?

tbrisker replied:
Looks like this is a bootstrap issue that occurs when opening modal when the scroll bar is visible: https://github.com/twbs/bootstrap/issues/9855

ohadlevy commented on app/views/smart_proxies/plugins/_puppet.html.erb:
```diff
@@ -0,0 +1,29 @@
+<%= show_feature_version('puppet') %>
+<div class="col-md-2"><%= _('Hosts managed:') %></div>
+<div class="col-md-10">
+  <%= link_to @hosts.length, hosts_path(:search => "puppetmaster = #{@smart_proxy.name}") %>
```
if the puppet master name has spaces in it, you would need to escape quotas, this is true for all other searches that we already included before /cc @shlomizadok

tbrisker replied:
fixed

ohadlevy commented:
also, the spacing between the secondary tabs and content imho is too small
![recorded](https://cloud.githubusercontent.com/assets/20167/12376924/8f59ede4-bd13-11e5-912e-661a28f32f63.gif)

ohadlevy commented:
another thing, not sure if its related, but other version checks stopped working after this patch
![recorded](https://cloud.githubusercontent.com/assets/20167/12376928/ba687c8a-bd13-11e5-846b-cba97405afcd.gif)

ohadlevy commented on app/services/smart_proxies/puppet_ca.rb:
```diff
@@ -20,7 +20,6 @@ def all(proxy)
         end.compact
 
-        # save our CA details for 5 seconds
         Rails.cache.write("ca_#{proxy.id}", certs, {:expires_in  => 1.minute }) if Rails.env.production?
```
TODO: move this to the status object as well?

tbrisker replied:
Shouldn't all these class methods move to a status object?

ohadlevy commented on app/helpers/puppetca_helper.rb:
```diff
@@ -1,8 +1,22 @@
+  STATES = [N_('pending'), N_('valid'), N_('revoked')]
+  CA_ICONS = { 'valid' => 'pficon pficon-ok',
+               'pending' => 'pficon pficon-warning-triangle-o',
+               'revoked' => 'pficon pficon-error-circle-o' }
```
I'm not sure revoked should be shown as error, its mostly FYI (as revoked is normal operation).

ohadlevy replied:
unresolved yet.

tbrisker replied:
changed it to `fa-ban`

ohadlevy commented on app/views/common/_card_status_mini.html.erb:
```diff
@@ -0,0 +1,15 @@
+<div class="<%= size %>">
+  <div class="card-pf card-pf-accented card-pf-aggregate-status card-pf-aggregate-status-mini">
+    <h2 class="card-pf-title">
+      <span class="<%= icon %>"></span>
```
not sure why the icon is to the left, imho they could be next to the text/counter.

tbrisker replied:
This is the patternfly layout for cards

lzap commented:
I can help with testing.

dLobatog commented on app/views/smart_proxies/plugins/_puppet_envs.html.erb:
```diff
@@ -0,0 +1,20 @@
+  <tbody>
+    <% envs.each do |name, count| %>
+      <tr>
+        <td class="ellipsis"><%= name %></td>
```
Link to the environment in Foreman

tbrisker replied:
The environment doesn't have a show page, and the form only allows changing the name (which won't affect the proxy anyways) and associating with taxonomies, I don't think this is useful.

dLobatog commented on app/views/smart_proxies/plugins/_puppet_envs.html.erb:
```diff
@@ -0,0 +1,20 @@
+    <% end %>
+    <tfoot>
+      <th><%= _("Total") %></th>
+      <th><%= envs.values.reduce(&:+) %></th>
```
Link to the puppet classes associated with that environment

tbrisker replied:
added (on the count above)

dLobatog commented on app/views/smart_proxies/plugins/_puppet_envs.html.erb:
```diff
@@ -0,0 +1,20 @@
+<% if envs.present? %>
```
Render an `empty_state`, for instance the partial `environments/help` when envs is not present too. I think it'd be clearer to do this at the controller level though.

tbrisker replied:
added empty state here, the partial for environments is not relevant as no envs indicates there is an issue with the puppetmaster setup.

dLobatog commented on lib/proxy_api/puppet.rb:
```diff
@@ -27,6 +27,16 @@ def classes(env)
+      pcs = parse(get "environments/#{env}/classes")
+      pcs.length
+    rescue RestClient::ResourceNotFound
+      0
```
When is this exception thrown? It looks like it could be hiding deeper problems

ohadlevy replied:
well, it could be that a certain puppet env is not synced to the specific proxy?

tbrisker replied:
Removing this, this should not occur and certainly shouldn't return 0 if it does.

dLobatog commented on app/views/smart_proxies/plugins/_puppet.html.erb:
```diff
@@ -0,0 +1,29 @@
+<div class="col-md-10">
+  <%= link_to @hosts.length, hosts_path(:search => "puppetmaster = \"#{@smart_proxy.name}\"") %>
+</div>
+<ul id="proxy-puppet-tab" class="nav nav-tabs nav-tabs-pf">
```
Double tab navigation (we have tabs for features + tabs for reports & statistics) is usually known as an UI antipattern. How about showing all sections, with 3 headers? General, Environments, Activity

ohadlevy replied:
@dLobatog we were actually following patternfly pattern -> https://rawgit.com/patternfly/patternfly/master/tests/tab.html#

dLobatog replied:
Right, secondary tabs are fine, this is like another tabbed context within a tab, notice how secondary tabs are right below the tabs on the Patternfly example

tbrisker replied:
I'm not how to handle this, I think this looks better and is very intuitive to use, where a long page with headers will look worse in my opinion (and would not work for the puppetca anyways because of the two tables of changing sizes.).

ohadlevy replied:
I wonder if we should introduce a primary tab instead of showing the
version above the secondary tabs.

dLobatog commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -45,6 +45,12 @@ def tftp_server
+  def puppet_environments
+    render :partial => 'smart_proxies/plugins/puppet_envs', :locals => {:envs => @smart_proxy.environment_stats}
+  rescue Foreman::Exception => exception
+    process_ajax_error exception
```
The messages rendered when the proxy isn't reachable don't look great - 
![screenshot from 2016-01-18 21-19-59](https://cloud.githubusercontent.com/assets/598891/12402566/282aa1b0-be2e-11e5-92f4-d5ec76cc3991.png)

Same goes for PuppetCA & Autosign

tbrisker replied:
fixed

ohadlevy commented on app/views/smart_proxies/plugins/_puppet_ca.html.erb:
```diff
@@ -0,0 +1,27 @@
+<h3><%= _("Puppet Certificate Authority") %></h3>
+<%= show_feature_version('puppetca') %>
```
can you please show when the CA expires ? (the CA is the first certificate - oldest)
we can probably highlight (maybe in a separate PR or not) certs that will expire - but IMHO the CA cert will always expire first (and revoke all the other certs along) so other certificate expiry does not really matter.

tbrisker replied:
Would this be the certificate that matches the proxy's hostname? if so, added.

ohadlevy replied:
I've added http://projects.theforeman.org/issues/12948 for a robust way of knowing the ca info, but however, for now, a simple sort for oldest cert would suffice (as the CA is the first certificate created).

lzap commented on app/controllers/autosign_controller.rb:
```diff
@@ -3,13 +3,16 @@ def index
 
-    autosign = @api.autosign
-    @autosign = autosign.paginate :page => params[:page], :per_page => Setting::General.entries_per_page
+    @autosign = @api.autosign
```
Are you giving up pagination completely? This will definitely break the page. Can you make sure the `Setting::General.entries_per_page` global setting is used? Mabe we should introduce new variable `Setting::General.entries_per_embedded_page` for inlined tables which will perhaps need smaller number than ordinary index pages.

ohadlevy replied:
@lzap pagination is now done via js with data tables

lzap replied:
And is there a configuration value for no of rows? Or is this automatic detection?

dLobatog replied:
It can be configured using lengthMenu - https://datatables.net/examples/advanced_init/length_menu.html

tbrisker replied:
@dLobatog we are currently using the legacy version of datatables. 
@lzap right now this uses the default which is 10, and I feel this fits nicely on this page. Do you think this should be changed?

dLobatog commented on app/controllers/autosign_controller.rb:
```diff
@@ -17,9 +20,9 @@ def create
 
     if @api.set_autosign(params[:id])
-      process_success({:success_redirect => smart_proxy_autosign_index_path(@proxy), :object_name => 'puppet autosign entry'})
+      process_success({:success_redirect => smart_proxy_path(@proxy, :anchor => 'autosign'), :object_name => 'puppet autosign entry'})
```
`create` and `delete` are the same, please consider extracting the common behavior to a shared method.

tbrisker replied:
done

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -26,10 +26,24 @@ def tftp_server
     end
   end
 
+  # TODO: extract to another status implementation
```
It looks like status is going to be merged today, so maybe this todo has to be completed

dLobatog commented on app/services/proxy_status.rb:
```diff
@@ -26,10 +26,24 @@ def tftp_server
     # As memcached does not support delete_matched, we need to delete each
     Rails.cache.delete(versions_cache_key)
     Rails.cache.delete("proxy_#{proxy.id}/tftp_server")
+    Rails.cache.delete("proxy_#{proxy.id}/environments")
```
`puppet_environments` rather, just so we can have more config management system environments

tbrisker replied:
changed

dLobatog commented:
@tbrisker Thanks for taking over this! A list of stuff that's pending:
- [ ] Find a way to present environment count in a more compelling way (@ohadlevy - i'm fine with the way it is now) - https://github.com/theforeman/foreman/pull/3018#discussion_r49725031 
- [ ] Move environments to its own ProxyStatus - https://github.com/theforeman/foreman/pull/3018#discussion_r49945570 , https://github.com/theforeman/foreman/pull/3069
- [ ] Link environment title to environment in Foreman - https://github.com/theforeman/foreman/pull/3018#discussion_r50037830
- [ ] Link puppet classes in environment with puppet classes in Foreman - https://github.com/theforeman/foreman/pull/3018#discussion_r50037857
- [ ] Render an empty state when there are no environments in the smart proxy - https://github.com/theforeman/foreman/pull/3018#discussion_r50038057
- [ ] Show user what was the failure when the class count could not be retrieved from the proxy - https://github.com/theforeman/foreman/pull/3018#discussion_r50038257
- [ ] Make failures when the proxy isn't available, or when the puppetmaster service is off look nicer - https://github.com/theforeman/foreman/pull/3018#discussion_r50040337
- [ ] Show CA expiration - https://github.com/theforeman/foreman/pull/3018#discussion_r50079024
- [ ] DRY create/delete in autosign controller - https://github.com/theforeman/foreman/pull/3018#discussion_r50092538
- [ ] Use `puppet_environments` cache key instead of `environments` to ensure it won't conflict with other config management systems - https://github.com/theforeman/foreman/pull/3018#discussion_r50093654

lzap commented on app/views/dashboard/_reports_widget.html.erb:
```diff
@@ -14,13 +14,13 @@
-          <td><%= report_event_column(report.failed_restarts, "label-warning") %></td>
-          <td><%= report_event_column(report.skipped, "label-info") %></td>
-          <td><%= report_event_column(report.pending, "label-info") %></td>
+          <td class='ellipsis'><%= link_to report.host, host_config_reports_path(report.host) %></td>
```
Ellipsis does not seem to work correctly, in my table it simply overflows. Is there a better solution?

tbrisker replied:
This is very strange, does it work for you on other tables? what browser are you on?

lzap replied:
Chrome stable, here is how it looks like in my case:

![image](https://cloud.githubusercontent.com/assets/49752/12425573/acbf35ec-bed5-11e5-84a9-22f5c4fe1d5c.png)

Here is the content that does not fit:

```
Certificate: Data: Version: 3 (0x2) Serial Number: 2 (0x2) Signature Algorithm: sha256WithRSAEncryption Issuer: CN=Puppet CA: lzapx.brq.redhat.com Validity Not Before: Jul 10 13:21:45 2014 GMT Not After : Jul 10 13:21:45 2019 GMT Subject: CN=lzapx.brq.redhat.com Subject Public Key Info: Public Key Algorithm: rsaEncryption Public-Key: (4096 bit) Modulus: 00:d7:84:99:fe:fc:5b:82:f9:5b:0c:e8:83:a1:65: 5e:b8:55:60:28:ab:49:51:6f:36:f5:32:eb:79:7f: 03:74:bf:01:70:70:80:ea:93:71:83:85:19:3f:ef: bc:de:0f:e6:84:52:35:d3:53:ac:ce:c5:45:41:3d: 72:60:37:09:af:95:d1:0f:e7:db:1f:96:ef:a4:98: 16:d7:d7:90:5a:4d:76:21:00:b7:65:02:75:d2:e1: 1c:64:ec:2e:44:6f:22:64:c6:38:73:06:35:07:cc: 1d:39:14:3b:72:29:53:b2:14:77:39:62:42:e3:21: c1:78:62:b5:3f:80:02:57:32:53:cb:84:d0:99:7e: 4a:7c:07:79:f7:52:b4:35:59:40:a1:bf:7f:b0:14: d5:8b:ab:7b:91:3a:de:4a:8e:16:ed:8d:7b:db:2f: eb:a4:15:6a:84:1f:8c:e5:db:1d:a4:79:e3:c6:77: 1d:62:79:28:e3:43:54:1e:d9:29:3c:f5:28:62:17: 0d:a7:03:ff:97:4e:01:e9:74:36:32:dd:7a:c2:a3: 99:53:9c:a1:c7:e2:ef:db:42:65:a6:59:1d:f7:71: c8:07:7c:03:02:80:e8:ea:85:34:0b:f1:b8:2e:5c: 90:88:34:13:b8:36:de:83:c9:ef:b7:a9:e8:65:57: 2d:99:a9:b5:d2:3e:0d:b5:0d:e4:82:e8:3a:fd:45: c4:92:08:88:ca:b1:58:6a:8c:b3:87:76:e7:27:60: da:bc:c9:d0:12:40:67:b0:bf:8b:87:db:54:b0:c0: 31:ae:25:bb:cb:07:31:be:54:f2:93:96:15:78:b3: dc:7c:c9:e2:93:e6:a8:91:48:ca:d3:1b:da:10:b9: 5c:57:d2:64:2d:3a:29:44:bf:d7:66:79:3b:e1:f9: 8a:58:61:ba:8f:94:38:c8:e6:e8:28:96:a9:c9:1f: 30:f1:c9:af:59:2a:0a:60:1e:89:e1:74:6f:4e:52: 93:23:04:38:aa:81:8d:d7:e9:0f:02:f4:9d:04:d7: 54:39:48:3e:4a:7d:9a:5e:aa:92:1b:cb:94:36:31: eb:65:f3:a4:6c:de:a0:ed:19:9f:e7:8c:2f:a3:b1: 19:93:ad:ca:20:45:0e:e0:8f:8f:ec:7e:ff:d2:d0: 9b:f4:19:20:f1:59:e2:c7:95:e3:4b:45:51:b1:3c: eb:84:2f:60:66:ab:5a:ef:df:52:6a:1d:94:bd:54: 6b:ec:e9:69:7b:44:7b:3c:90:1a:af:bb:b0:7f:31: b8:fe:73:83:b1:97:ff:4f:8f:70:03:78:68:fc:d6: 73:46:60:f3:4b:a1:f8:61:45:d4:ae:f1:bc:3b:6f: 1e:fb:23 Exponent: 65537 (0x10001) X509v3 extensions: Netscape Comment: Puppet Ruby/OpenSSL Internal Certificate X509v3 Subject Alternative Name: DNS:lzapx.brq.redhat.com, DNS:puppet, DNS:puppet.brq.redhat.com X509v3 Key Usage: critical Digital Signature, Key Encipherment X509v3 Extended Key Usage: critical TLS Web Server Authentication, TLS Web Client Authentication X509v3 Basic Constraints: critical CA:FALSE X509v3 Subject Key Identifier: 58:FB:B9:7F:ED:AF:F8:6F:45:9D:78:BA:A6:F9:91:EE:7C:16:1A:35 X509v3 Authority Key Identifier: keyid:11:35:7C:0A:23:4D:3C:26:16:49:39:9F:4D:57:CE:58:C9:3C:4F:F8 Signature Algorithm: sha256WithRSAEncryption 59:3b:36:3a:7b:7e:a7:c8:fb:06:89:d7:25:5f:a9:dc:f6:68: 96:a8:f2:81:fb:4f:aa:38:9f:26:67:75:f3:4d:c6:db:44:4d: e8:58:9d:23:b0:73:c4:dc:d3:04:db:e4:20:f9:a2:65:14:38: 2d:fa:77:ef:63:4f:56:1e:1c:e7:61:db:dc:f5:9b:68:2f:3a: 9f:b7:a8:b5:66:b2:84:79:db:ac:ae:89:7c:d0:86:bd:41:cd: 2e:17:02:29:d2:73:95:75:29:fb:39:c4:f2:d5:82:2a:3e:6a: 19:e8:ee:e0:b3:4b:fe:af:03:44:28:bf:f5:01:b8:ac:22:ec: c4:8e:55:81:14:1d:e0:c4:45:c5:5a:e3:42:f0:fe:12:1d:2b: 77:0d:f7:35:26:91:c0:b5:3d:7b:b7:3a:dd:df:9a:24:1c:ef: e6:05:c4:4b:e9:cf:c6:73:b9:43:a1:22:5c:a5:8a:7e:fe:64: 35:f0:77:94:2f:73:76:27:08:db:35:d0:05:28:9d:47:8b:9d: dd:dc:7c:e3:b8:8d:5b:d8:04:1c:c1:f6:26:b5:76:f2:85:1b: 23:51:1c:40:0c:cf:2b:82:3f:a9:94:35:98:06:6a:35:5b:54: 76:3d:df:f8:f0:be:06:a7:61:7e:c4:15:bb:95:7d:75:bd:68: 18:c2:a2:c6:91:db:b8:25:ea:7f:56:76:3d:10:bd:ea:3a:66: c4:25:1f:93:f0:85:7c:ae:ff:44:6c:3c:a7:b6:30:d8:d5:28: 24:39:09:03:01:1c:85:73:c7:cb:b7:66:1c:5e:59:9a:f1:a8: 86:7c:d4:fb:12:58:a0:1e:89:c4:2f:c6:e9:70:81:9f:d5:cf: 2d:c3:87:42:bb:05:95:0e:e3:1c:3b:d2:cf:ca:99:d3:16:e5: ae:0d:e8:5d:48:f9:7c:d7:7a:bd:74:5d:4c:64:0a:0f:be:18: 33:aa:06:07:52:b0:22:7f:a7:11:1e:6b:11:05:b7:ff:c6:04: c8:91:7b:44:78:10:cd:76:31:20:84:9d:f8:23:1a:37:52:8a: 54:32:ec:a0:d4:15:a5:5e:38:9c:67:d5:41:bf:d1:72:16:6d: eb:ff:7e:f4:14:01:b8:1a:2c:95:6a:5e:68:6f:07:1c:eb:8d: ce:bc:09:bf:6a:de:67:a6:1a:73:ef:dd:64:3a:47:7d:e1:4b: ec:a9:ea:97:9d:b5:d4:38:a4:0b:1a:79:b1:74:a7:59:d1:43: 13:77:bc:e6:66:c6:e2:70:4a:c1:29:4d:fe:a4:69:4b:34:ae: 45:d2:d6:ce:01:f4:3a:51:86:81:70:03:f7:0b:ee:e6:6b:f7: 2e:cc:b0:0b:80:90:8f:d1
```

lzap replied:
Oh I was missing `table-fixed reports-table`. It's okay now.

ohadlevy replied:
@lzap please try avoiding pulling another PR review into this one, if you do wish to get someone attention link it from the other pr.

lzap commented on app/assets/javascripts/application.js:
```diff
@@ -136,6 +134,15 @@ function onContentLoad(){
+        "sPaginationType": "bootstrap"
+      }
+  );
+}
```
Looks like in my tab, I will need completely different datatable initialization, can you rename the key from `inline` to something more specific, so my table can use it's own key?

tbrisker replied:
this is currently used both by my PR and in the compute resource view. Would it be possible for you to use a different selector then `[data-table=inline]`?

lzap replied:
I will use different selector then, thanks.

tbrisker commented:
@dLobatog @ohadlevy Thanks for all the comments. 
I have either fixed or responded to all of them I believe, let me know if I missed something.
The only thing I didn't change yet in this update is the separation of the status objects which depends on #3069, I will fix it once that is merged.

ohadlevy commented:
Message from application: /home/foreman/git/foreman/app/controllers/autosign_controller.rb:40: syntax error, unexpected end-of-input, expecting keyword_end (SyntaxError)

tbrisker commented:
@ohadlevy sorry, missed an `end` and pushed again a minuter later, please pull again

ohadlevy commented:
@tbrisker please test again...

```
undefined method `expires_at' for nil:NilClass
/home/foreman/git/foreman/app/views/smart_proxies/plugins/_puppet_ca.html.erb:9:in `_cf744db30a28ac5da9b698091f0d0fc9'
/home/foreman/gems/gems/actionview-4.1.5/lib/action_view/template.rb:145:in `block in render'
/home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/notifications.rb:161:in `instrument'
/home/foreman/gems/gems/actionview-4.1.5/lib/action_view/template.rb:339:in `instrument'
/home/foreman/gems/gems/actionview-4.1.5/lib/action_view/template.rb:143:in `render'
/home/foreman/gems/gems/deface-1.0.2/lib/deface/action_view_extensions.rb:41:in `render'
/home/foreman/gems/gems/actionview-4.1.5/lib/action_view/renderer/partial_renderer.rb:306:in `render_partial'
/home/foreman/gems/gems/actionview-4.1.5/lib/action_view/renderer/partial_renderer.rb:279:in `block in render'
/home/foreman/gems/gems/actionview-4.1.5/lib/action_view/renderer/abstract_renderer.rb:38:in `block in instrument'
/home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/notifications.rb:159:in `block in instrument'
/home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/notifications/instrumenter.rb:20:in `instrum
```

ohadlevy commented on app/views/smart_proxies/plugins/_dhcp.html.erb:
```diff
@@ -1,10 +1,10 @@
 <div class="row">
   <h3><%= _('DHCP') %></h3>
```
this should be above the row class

ohadlevy commented on app/views/smart_proxies/plugins/_puppet.html.erb:
```diff
@@ -0,0 +1,29 @@
+<% dashboard = Dashboard::Data.new("puppetmaster = #{@smart_proxy.name}") %>
```
please convert the dashboard data fetching into a separate ajax calls.

ohadlevy replied:
also note, this generates many expensive SQL queries

``` sql
(6813.3ms)  SELECT COUNT(DISTINCT "hosts"."id") FROM "hosts" LEFT OUTER JOIN "smart_proxies" ON "smart_proxies"."id" = "hosts"."puppet_proxy_id" INNER JOIN reports ON reports.host_id = hosts.id WHERE "hosts"."type" IN ('Host::Managed') AND (("smart_proxies"."name" = 'foreman')) AND (reports.reported_at BETWEEN '2016-01-20 13:34:54.933765' AND '2016-01-20 13:37:54.933765')
```

tbrisker commented:
@dLobatog @ohadlevy I have refactored this to use the new proxy status object structure. 
I have also (I believe) addressed all of your comments so far. 
Let me know if there are any more issues which must be handled before this can be merged.
As this is getting to be quite a large PR, I would prefer if any issues that aren't crucial at this point and can be handled separately could be delayed to further iterations.

dLobatog commented:
:+1: - I don't have any major comments at first glance. I'll wait for @ohadlevy to ack, but in my opinion it's ready to merge.

dLobatog commented:
A couple of minor comments actually, secondary tabs don't look how they should look like on Patternfly (https://www.patternfly.org/widgets/#tabs), also some tests to cover the PuppetCA status would help too.

ohadlevy commented:
@tbrisker couple of visual comments about failures:
![proxy-puppet-failure](https://cloud.githubusercontent.com/assets/20167/12462565/ab511aca-bfc7-11e5-938a-5cd80875a631.png)
![puppetca-spinners](https://cloud.githubusercontent.com/assets/20167/12462566/ab53b532-bfc7-11e5-8b94-8a158bb83fb0.png)

not related, but we might need to disable the tabs and other ajax calls if the initial ping failed.

ohadlevy commented:
also, we should probably not cache the result if there was an exception

tbrisker commented:
@ohadlevy the three spinners and error messages are because each card is loaded separately. Theoretically, there could be a situation when only some of them load.
Disabling all ajax calls when the ping fails might be problematic since they are all fired asynchronously when the page loads, I would not want to wait for the ping response before firing the rest as it would slow down the entire loading time in case the ping is slow.
Exceptions are not cached, see `ProxyStatus::Base#fetch_proxy_data`.

ohadlevy commented on app/services/proxy_status/base.rb:
```diff
@@ -22,15 +22,19 @@ def self.humanized_name
 
     attr_reader :proxy, :cache_duration, :cache
 
+    def api_class
```
nitpick: should be protected?

tbrisker replied:
fixed

ohadlevy commented on app/services/smart_proxies/puppet_ca_certificate.rb:
```diff
@@ -0,0 +1,30 @@
+require "time"
+
+class SmartProxies::PuppetCACertificate
+  attr_reader :name, :state, :fingerprint, :valid_from, :expires_at, :smart_proxy_id
```
imho you wont need to pass the proxy id, rather then to link to the proxy status object, this will you to delegate the sign / delete etc actions to the ca / status object instead see https://github.com/ohadlevy/foreman/commit/3422a9105199dd6f78a90f9a9a4cf77f4dcd1c10#diff-3bc418a60798c8ad4833276ef193c557

tbrisker replied:
fixed, thank you

ohadlevy commented on app/views/autosign/_counts.html.erb:
```diff
@@ -0,0 +1,22 @@
+  <div class="card-pf  card-pf-accented card-pf-aggregate-status card-pf-with-action">
+    <h2 class="card-pf-title">
+      <a href="#autosign">
+        <%= icon_text('check', '', :kind => 'fa') %><span class="card-pf-aggregate-status-count"><%= autosign.length.to_s %></span> <%= _("Autosign entries") %>
```
does this take into account user permissions?

tbrisker replied:
this card is only displayed for users with the `:view_smart_proxies_autosign` permission

ohadlevy commented on app/services/proxy_status/puppetca.rb:
```diff
@@ -0,0 +1,59 @@
+    end
+
+    def expiry
+      ca_cert = find(@proxy.hostname)
```
this will not work in all / most cases. IMHO just sort based on the oldest cert.

tbrisker replied:
fixed to sort certs on creation time.

dLobatog commented on app/services/proxy_status/puppetca.rb:
```diff
@@ -0,0 +1,71 @@
+  class PuppetCA < Base
+    def certs
+      fetch_proxy_data('/certs') do
+        certs = api.all.map do |name, properties|
```
unused variable

tbrisker replied:
removed, leftover from debugging

dLobatog commented on test/unit/proxy_statuses/proxy_status_puppetca_test.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+
+  test 'it returns autosign' do
+    assert_equal('autosigned.com', @proxy_status.autosign[0])
```
Is this testing anything at all? autosign is stubbed above, maybe you want to `expect` a proxy call or something similar

tbrisker replied:
changed to expect, even though the stub was on the api, not the status.

dLobatog commented on test/unit/proxy_statuses/proxy_status_puppetca_test.rb:
```diff
@@ -0,0 +1,31 @@
+    @proxy_status = ProxyStatus::PuppetCA.new(@proxy)
+  end
+
+  test 'it returns CA certificates' do
```
It's missing a test for the case where there's no CA certificate

tbrisker replied:
added

dLobatog commented on app/services/proxy_status/puppetca.rb:
```diff
@@ -0,0 +1,71 @@
+
+    def destroy(cert)
+      Rails.cache.delete(cache_key+'/certs')
+      api.del_certificate(cert)
```
You probably want to test these methods (sign, destroy, autosign, set_autosign, del_autosign) via expectations, to make sure we're making ProxyAPI::PuppetCA calls for them.

tbrisker replied:
added

lzap commented:
Nitpick: Since both version and logs are AJAX only and using V2 API, I wonder if you could do the same - you could get rid of UI actions completely and simply use the V2 API.

tbrisker commented:
@lzap Not sure what you mean by this, is this related to this PR or a general statement about the proxy status objects?

tbrisker commented:
@dLobatog Added more tests.

ohadlevy commented on app/services/proxy_status/puppetca.rb:
```diff
@@ -0,0 +1,71 @@
+    end
+
+    def sign(cert)
+      Rails.cache.delete(cache_key+'/certs')
```
use revoke_cache! instead?

tbrisker replied:
I don't need to delete the autosign cache when changing the certificates data and vice versa

ohadlevy replied:
i understand, the main thing that bother me here is that you communicate directly with the "cache" vs thought the base class.
maybe instead we should keep a cache method on the base class

tbrisker replied:
How about adding an optional subkey to `revoke_cache!` on the base class and `super.revoke_cache!('/autosign')` from the child?

ohadlevy replied:
sounds good to me.

tbrisker replied:
done.

ohadlevy commented on app/services/proxy_status/puppetca.rb:
```diff
@@ -0,0 +1,71 @@
+    def revoke_cache!
+      Rails.cache.delete(cache_key+'/certs')
+      Rails.cache.delete(cache_key+'/autosign')
+      super
```
why super? i dont know if its any useful in this case?

tbrisker replied:
Right, it doesn't actually do anything as nothing is cached under just cache_key. removing.

ohadlevy commented:
@dLobatog :+1: from my side.

ohadlevy commented:
when revoking a cert i saw the following:
`Error: super: no superclass method `destroy' for #<ProxyStatus::PuppetCA:0x007f1859efe278>`

an additional trace showed:

```

2016-01-24T13:35:09 [app] [E] Marshalling error for key 'proxy_7/PuppetCA/certs': no _dump_data is defined for class OpenSSL::X509::Certificate
2016-01-24T13:35:09 [app] [E] You are trying to cache a Ruby object which cannot be serialized to memcached.
2016-01-24T13:35:09 [app] [E] /home/foreman/gems/gems/dalli-2.7.5/lib/dalli/server.rb:414:in `dump'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/dalli/server.rb:414:in `serialize'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/dalli/server.rb:286:in `set'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/dalli/server.rb:67:in `request'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/dalli/options.rb:18:in `block in request'
 |  /opt/rh/rh-ruby22/root/usr/share/ruby/monitor.rb:211:in `mon_synchronize'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/dalli/options.rb:17:in `request'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/dalli/client.rb:348:in `perform'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/dalli/client.rb:120:in `set'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/active_support/cache/dalli_store.rb:314:in `write_entry'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/cache/strategy/local_cache.rb:115:in `write_entry'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/active_support/cache/dalli_store.rb:144:in `block (2 levels) in write'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/dalli/client.rb:256:in `with'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/active_support/cache/dalli_store.rb:82:in `with'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/active_support/cache/dalli_store.rb:142:in `block in write'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/active_support/cache/dalli_store.rb:370:in `block in instrument'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/notifications.rb:161:in `instrument'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/active_support/cache/dalli_store.rb:370:in `instrument'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/active_support/cache/dalli_store.rb:141:in `write'
 |  /home/foreman/gems/gems/dalli-2.7.5/lib/active_support/cache/dalli_store.rb:115:in `fetch'
 |  /home/foreman/git/foreman/app/services/proxy_status/base.rb:39:in `fetch_proxy_data'
 |  /home/foreman/git/foreman/app/services/proxy_status/puppetca.rb:4:in `certs'
 |  /home/foreman/git/foreman/app/controllers/puppetca_controller.rb:12:in `counts'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_controller/metal/implicit_render.rb:4:in `send_action'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/abstract_controller/base.rb:189:in `process_action'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_controller/metal/rendering.rb:10:in `process_action'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/abstract_controller/callbacks.rb:20:in `block in process_action'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:113:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:113:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:299:in `block (2 levels) in halting'
 |  /home/foreman/gems/gems/rails-observers-0.1.2/lib/rails/observers/action_controller/caching/sweeping.rb:73:in `around'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:447:in `public_send'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:447:in `block in make_lambda'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:298:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:298:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:299:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:299:in `block (2 levels) in halting'
 |  /home/foreman/git/foreman/app/controllers/concerns/application_shared.rb:13:in `set_timezone'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:424:in `block in make_lambda'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:298:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:298:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:149:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:149:in `block in halting_and_conditional'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:149:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:149:in `block in halting_and_conditional'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:149:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:149:in `block in halting_and_conditional'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:229:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:229:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:149:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:149:in `block in halting_and_conditional'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:299:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:299:in `block (2 levels) in halting'
 |  /home/foreman/git/foreman/app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:424:in `block in make_lambda'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:298:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:298:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:229:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:229:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:299:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:299:in `block (2 levels) in halting'
 |  /home/foreman/gems/gems/rails-observers-0.1.2/lib/rails/observers/action_controller/caching/sweeping.rb:73:in `around'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:447:in `public_send'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:447:in `block in make_lambda'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:298:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:298:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:166:in `block in halting'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:86:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:86:in `run_callbacks'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/abstract_controller/callbacks.rb:19:in `process_action'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_controller/metal/rescue.rb:29:in `process_action'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_controller/metal/instrumentation.rb:31:in `block in process_action'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/notifications.rb:159:in `block in instrument'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/notifications.rb:159:in `instrument'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_controller/metal/instrumentation.rb:30:in `process_action'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_controller/metal/params_wrapper.rb:250:in `process_action'
 |  /home/foreman/gems/gems/activerecord-4.1.5/lib/active_record/railties/controller_runtime.rb:18:in `process_action'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/abstract_controller/base.rb:136:in `process'
 |  /home/foreman/gems/gems/actionview-4.1.5/lib/action_view/rendering.rb:30:in `process'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_controller/metal.rb:196:in `dispatch'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_controller/metal/rack_delegation.rb:13:in `dispatch'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_controller/metal.rb:232:in `block in action'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/routing/route_set.rb:82:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/routing/route_set.rb:82:in `dispatch'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/routing/route_set.rb:50:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/journey/router.rb:71:in `block in call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/journey/router.rb:59:in `each'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/journey/router.rb:59:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/routing/route_set.rb:678:in `call'
 |  /home/foreman/gems/gems/apipie-rails-0.3.5/lib/apipie/static_dispatcher.rb:65:in `call'
 |  /home/foreman/gems/gems/apipie-rails-0.3.5/lib/apipie/extractor/recorder.rb:132:in `call'
 |  /home/foreman/gems/gems/apipie-rails-0.3.5/lib/apipie/middleware/checksum_in_headers.rb:27:in `call'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/etag.rb:23:in `call'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/conditionalget.rb:25:in `call'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/head.rb:11:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/middleware/params_parser.rb:27:in `call'
 |  /home/foreman/git/foreman/lib/middleware/catch_json_parse_errors.rb:9:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/middleware/flash.rb:254:in `call'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/session/abstract/id.rb:225:in `context'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/session/abstract/id.rb:220:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/middleware/cookies.rb:560:in `call'
 |  /home/foreman/gems/gems/activerecord-4.1.5/lib/active_record/query_cache.rb:36:in `call'
 |  /home/foreman/gems/gems/activerecord-4.1.5/lib/active_record/connection_adapters/abstract/connection_pool.rb:621:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/callbacks.rb:82:in `run_callbacks'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/middleware/callbacks.rb:27:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/middleware/remote_ip.rb:76:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
 |  /home/foreman/gems/gems/railties-4.1.5/lib/rails/rack/logger.rb:38:in `call_app'
 |  /home/foreman/gems/gems/railties-4.1.5/lib/rails/rack/logger.rb:22:in `call'
 |  /home/foreman/gems/gems/quiet_assets-1.1.0/lib/quiet_assets.rb:27:in `call_with_quiet_assets'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/middleware/request_id.rb:21:in `call'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/methodoverride.rb:21:in `call'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/runtime.rb:17:in `call'
 |  /home/foreman/gems/gems/activesupport-4.1.5/lib/active_support/cache/strategy/local_cache_middleware.rb:26:in `call'
 |  /home/foreman/gems/gems/actionpack-4.1.5/lib/action_dispatch/middleware/static.rb:64:in `call'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/sendfile.rb:112:in `call'
 |  /home/foreman/gems/gems/railties-4.1.5/lib/rails/engine.rb:514:in `call'
 |  /home/foreman/gems/gems/railties-4.1.5/lib/rails/application.rb:144:in `call'
 |  /home/foreman/gems/gems/railties-4.1.5/lib/rails/railtie.rb:194:in `public_send'
 |  /home/foreman/gems/gems/railties-4.1.5/lib/rails/railtie.rb:194:in `method_missing'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/builder.rb:138:in `call'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/urlmap.rb:65:in `block in call'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/urlmap.rb:50:in `each'
 |  /home/foreman/gems/gems/rack-1.5.5/lib/rack/urlmap.rb:50:in `call'
 |  /root/.gem/gems/passenger-5.0.7/lib/phusion_passenger/rack/thread_handler_extension.rb:89:in `process_request'
 |  /root/.gem/gems/passenger-5.0.7/lib/phusion_passenger/request_handler/thread_handler.rb:149:in `accept_and_process_next_request'
 |  /root/.gem/gems/passenger-5.0.7/lib/phusion_passenger/request_handler/thread_handler.rb:110:in `main_loop'
 |  /root/.gem/gems/passenger-5.0.7/lib/phusion_passenger/request_handler.rb:414:in `block (3 levels) in start_threads'
 |  /root/.gem/gems/passenger-5.0.7/lib/phusion_passenger/utils.rb:111:in `block in create_thread_and_abort_on_exception'
 |  /home/foreman/gems/gems/logging-2.0.0/lib/logging/diagnostic_context.rb:448:in `call'
 |  /home/foreman/gems/gems/logging-2.0.0/lib/logging/diagnostic_context.rb:448:in `block in create_with_logging_context'
2016-01-24T13:35:09 [app] [I]   Rendered puppetca/_counts.html.erb (2.5ms)
2016-01-24T13:35:09 [app] [I] Completed 200 OK in 1594ms (Views: 6.3ms | ActiveRecord: 3.1ms)
```

tbrisker commented:
@ohadlevy I think it should be fixed now.

dLobatog commented:
Merged as 2ac8d7badce0267c5b5fca4977c070b28b3f4f58, thanks @tbrisker!

dLobatog commented:
Also thanks @lzap and @ohadlevy for the reviews!
